### PR TITLE
[juju] Detect LP bug 1852502

### PR DESCRIPTION
**Pull request by nkshirsagar**
Resolves: #253

### CODE REVIEW

**Comment by nkshirsagar** [comment]
tested 

# ./hotsos.sh /home/nikhil/Downloads/sosreport-juju-1-00325262-2022-01-04-qulgfac
INFO: analysing sosreport /home/nikhil/Downloads/sosreport-juju-1-00325262-2022-01-04-qulgfac  
hotsos:
  version: development
  repo-info: e7f31b1
system:
  hostname: juju-1
  os: ubuntu bionic
  num-cpus: 8
  load: 1.23, 1.55, 1.72
  virtualisation: kvm
  rootfs: /dev/vda2       95595340 18917080  71779168  21% /
  unattended-upgrades: ENABLED
  date: Tuesday 04 January 2022 09:23:17 AM UTC
  sysctl-mismatch:
    kernel.core_pattern:
      actual: '|/usr/share/apport/apport %p %s %c %d %P %E'
      expected: '%h.%E.%p.core'
  potential-issues:
    SystemWarnings:
      - Unattended upgrades are enabled which can lead to uncontrolled changes to
        this environment. If maintenance windows are required please consider disabling
        unattended upgrades. (origin=system.auto_scenario_check)
    SysCtlWarnings:
      - host sysctl not consistent with contents of systcl.d (origin=system.checks)
juju:
  services:
    systemd:
      enabled:
        - juju-db
        - jujud-machine-0
    ps:
      - jujud (1)
      - mongod (1)
  version: 2.9.16
  machine: '0'
  units:
    local:
      - '-'
  bugs-detected:
    - desc: known mongodb bug identified - https://jira.mongodb.org/browse/TOOLS-1636
        Workaround is to pass --no-logs to juju create-backup. This is an issue only
        with Mongo 3. Mongo 4 does not have this issue. Upstream is working on migrating
        to Mongo 4 in the Juju 3.0 release.
      id: https://bugs.launchpad.net/bugs/1852502
      origin: juju.auto_bug_check
kernel:
  version: 4.15.0-163-generic
  boot: ro console=tty1 console=ttyS0
  cpu:
    smt: disabled
    cpufreq-scaling-governor: unknown
  memory-checks: no issues found

INFO: see --help for more options

File: defs/bugs/juju/juju_core.yaml [hunk 1]
```diff
@@ -17,4 +17,14 @@ input:
       post-series-upgrade failed, issues with unit {} - see LP bug for
       workaround.
     search-result-format-groups: [1]
-
+1852502:
+  input:
```
Comment (dosaboy) [inline defs/bugs/juju/juju_core.yaml:21]: There is a another way to do this which will allow the check to also work on a live host as well as sosreport:

input:
type: command
value: journalctl
meta:
kwargs:
unit: juju-db

**Comment by dosaboy** [comment]
Thanks for the patch @nkshirsagar. We now also have the ability to check the version of the package to see if it has the fix included or not (see https://github.com/canonical/hotsos/blob/89ed7fa06f0f6a082083d54730bde511c4057b2f/defs/bugs/openstack/barbican.yaml#L4) but this currently only supports apt packages (https://github.com/canonical/hotsos/blob/89ed7fa06f0f6a082083d54730bde511c4057b2f/core/ycheck/bugs.py#L119). I wonder if it would be possible to check the version here but if it requires implementing snap support we can do that separately. I have left a comment above for something to tweak and it would also be great to have a unit test to test this check.

**Comment by nkshirsagar** [comment]
Regarding the unit test, this is what I've been trying for the unit test - (here's the full patch as it is currently - https://pastebin.canonical.com/p/cDPPTJ3MvG/ ..  

seems like it does not detect my newly added bug in the bugs yaml file, through the YBugChecker() call in the unit test,  so the assertion fails.. and it only find the one defined earlier

so I end up getting an error during the build for tests like this

======================================================================
FAIL: test_detect_known_bugs (test_juju.TestJujuKnownBugs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/nikhil/HDD_MOUNT/Downloads/hotsos_fork/hotsos/.tox/py3/lib/python3.8/site-packages/mock/mock.py", line 1346, in patched
    return func(*newargs, **newkeywargs)
  File "/home/nikhil/HDD_MOUNT/Downloads/hotsos_fork/hotsos/tests/unit/test_juju.py", line 119, in test_detect_known_bugs
    self.assertEqual(known_bugs_utils._get_known_bugs(), expected)
AssertionError: {'bugs-detected': [{'desc': 'Unit unit-rabbitmq-server-2 faile[138 chars]t'}]} != {'bugs-detected': [{'id': 'https://bugs.launchpad.net/bugs/191[252 chars]k'}]}
  {'bugs-detected': [{'desc': 'Unit unit-rabbitmq-server-2 failed to start due '
                              'to members in relation 236 that cannot be '
                              'removed.',
                      'id': 'https://bugs.launchpad.net/bugs/1910958',
-                     'origin': 'juju.01part'}]}
?                                             ^^

+                     'origin': 'juju.01part'},
?                                             ^

+                    {'desc': 'known mongodb bug.',
+                     'id': 'https://bugs.launchpad.net/bugs/1852502',
+                     'origin': 'juju.auto_bug_check'}]}
-------------------- >> begin captured logging << --------------------
root: DEBUG: loading bugs definitions for plugin=juju

File: defs/bugs/juju/juju_core.yaml (hunk 1 above)
Comment (nkshirsagar) [reply defs/bugs/juju/juju_core.yaml:21]: made this change, thank you.

**Comment by nkshirsagar** [comment]
> Regarding the unit test, this is what I've been trying for the unit test - (here's the full patch as it is currently - https://pastebin.canonical.com/p/cDPPTJ3MvG/ ..
> 
> seems like it does not detect my newly added bug in the bugs yaml file, through the YBugChecker() call in the unit test, so the assertion fails.. and it only finds the one defined earlier
> 
> so I end up getting an error during the build for tests like this
> 
> ```
> FAIL: test_detect_known_bugs (test_juju.TestJujuKnownBugs)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/nikhil/HDD_MOUNT/Downloads/hotsos_fork/hotsos/.tox/py3/lib/python3.8/site-packages/mock/mock.py", line 1346, in patched
>     return func(*newargs, **newkeywargs)
>   File "/home/nikhil/HDD_MOUNT/Downloads/hotsos_fork/hotsos/tests/unit/test_juju.py", line 119, in test_detect_known_bugs
>     self.assertEqual(known_bugs_utils._get_known_bugs(), expected)
> AssertionError: {'bugs-detected': [{'desc': 'Unit unit-rabbitmq-server-2 faile[138 chars]t'}]} != {'bugs-detected': [{'id': 'https://bugs.launchpad.net/bugs/191[252 chars]k'}]}
>   {'bugs-detected': [{'desc': 'Unit unit-rabbitmq-server-2 failed to start due '
>                               'to members in relation 236 that cannot be '
>                               'removed.',
>                       'id': 'https://bugs.launchpad.net/bugs/1910958',
> -                     'origin': 'juju.01part'}]}
> ?                                             ^^
> 
> +                     'origin': 'juju.01part'},
> ?                                             ^
> 
> +                    {'desc': 'known mongodb bug.',
> +                     'id': 'https://bugs.launchpad.net/bugs/1852502',
> +                     'origin': 'juju.auto_bug_check'}]}
> -------------------- >> begin captured logging << --------------------
> root: DEBUG: loading bugs definitions for plugin=juju
> ```

Thanks for your help in fixing this failing unit test @dosaboy  - I've looked at the changes in the patch that you showed me - https://pastebin.canonical.com/p/WnJn8R49fN/ , and pushed the changes to the PR. Also tested again with a sos that has the issue.

**Review by dosaboy — APPROVED** [review]
lgtm!

**Pull request merged**